8_bit_pic.py

```diff
@@ -11,102 +37,352 @@
+                # Pseudo-op
+                pseudoop = m.group("pseudoop").upper()
+                operands = m.group("operands").split(",")
+                self._process_pseudoop(m.group("name"), pseudoop, operands, line_num)
+            elif m := self.OPCODE_RE.match(line):
+                # Opcode
+                opcode = m.group("opcode").upper()
+                operands = m.group("operands") or ""
+                operands = operands.split(",")
+                self._process_opcode(m.group("label"), opcode, operands, line_num, tokens)
+            else:
+                raise Exception(f"Syntax error on line {line_num}.")
+
+        logger.debug(tokens)
+
+        return tokens
 
     def assemble(self):
+        self.byte_code = bytearray()
+
+        for token in self.tokenize():
+            if code_bytes := token.encode(self):
+                self.byte_code.extend(code_bytes)
+
+        logger.debug(self.byte_code)
+        logger.debug(self.symbol_table)
+
         return self.byte_code
 
     def link(self):
         return self.byte_code
 
     def compile(self):
+        self.assemble()
+        self.link()
         return self.byte_code
 
 
-@dataclass
-class CPUFlags: ...
+def resolve_int_value(assembler: vm_types.GenericAssembler, value: str, mask: int) -> int:
+    """Resolve integer referene"""
+    if value in assembler.symbol_table["symbols"]:
+        # Symbol reference.
+        value = assembler.symbol_table["symbols"][value]
+    elif value in assembler.symbol_table["map"]:
+        # Address reference.
+        value = assembler.symbol_table["map"][value]
+    else:
+        # Literal.
+        pass
+    try:
+        int_value = int(value)
+    except ValueError:
+        # Try to handle hexadecimal literals with "h" suffix.
+        int_value = int(value.lower().rstrip("h"), 16)
+    int_value = int(int_value) & mask
+    return int_value
 
 
 @dataclass
-class CPURegisters: ...
-
+class PICCentralProcessingUnit(vm_types.GenericCentralProcessingUnit):
+    RAM: bytearray
+    INSTRUCTION_MAP: ClassVar[
+        MutableMapping[InstructionCodes, Callable[[Self, int], None]]
+    ] = {}
+    ic: int = 0
+    pc: int = 0
+    wreg: int = 0
 
-@dataclass
-class CentralProcessingUnit:
     @classmethod
     def register_instruction(cls, inst_code) -> vm_types.DecoratorCallable:
         def decorator(f):
+            cls.INSTRUCTION_MAP[inst_code] = f
             return f
-
         return decorator
 
     @property
     def word_in_bytes(self) -> int:
-        return 1
+        return 2
 
     @property
     def nible_in_bytes(self) -> int:
         raise NotImplementedError
 
     def fetch(self):
-        raise NotImplementedError
+        next_ip = self.pc + self.word_in_bytes
+        self.ic = int.from_bytes(self.RAM[self.pc: next_ip])
+        self.pc = next_ip
 
     def reset(self):
-        return
+        self.ic = 0
+        self.pc = 0
+        self.wreg = 0
```

Comment:
I'll move these to `RAM` eventually and add the status and "top-of-stack registers.